### PR TITLE
Moved expires to resource metadata for services.Users.

### PR DESCRIPTION
**Pull request by russjones**
**Description**

Moved expiry field from spec to metadata for services.Users and updated
expiry check to prefer metadata and fallback to spec if not found. Added
test coverage.

### CODE REVIEW

File: lib/services/suite/suite.go [hunk 1]
```diff
@@ -41,6 +42,8 @@ import (
 	"gopkg.in/check.v1"
 )
 
+var _ = fmt.Printf
```
Comment (r0mant) [inline lib/services/suite/suite.go]: Need this?

File: lib/auth/saml_test.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	"gopkg.in/check.v1"
+)
+
+type SAMLSuite struct {
+	a *AuthServer
+	b backend.Backend
+	c clockwork.FakeClock
+}
+
+var _ = fmt.Printf
```
Comment (r0mant) [inline lib/auth/saml_test.go]: Need this?

**Review by r0mant — APPROVED** [review]

**Comment by klizhentas** [comment]
I think this problem applies to all users, not just github ones

**Review by klizhentas — CHANGES_REQUESTED** [review]
Can you please check that if this problem applies to all external types

**Comment by russjones** [comment]
@klizhentas It does and I fixed it for all external identity providers. I've fixed it for all and covered all with tests. That's what the bulk of the PR was.

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Comment by klizhentas** [comment]
k, sorry the PR only showed me two github files, let me re-review

**Review by klizhentas — APPROVED** [review]

**Comment by klizhentas** [comment]
retest this please

File: lib/auth/github.go [hunk 3]
```diff
@@ -210,61 +203,157 @@ func (s *AuthServer) validateGithubAuthCallback(q url.Values) (*GithubAuthRespon
...
+		compatibility: compatibility,
+	})
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return certs.ssh, certs.tls, nil
+}
+
+// createUserParams is a set of parameters used to create a user for an\
```
Comment (klizhentas) [inline lib/auth/github.go]: extra `\`?

File: lib/auth/github.go [hunk 4]
```diff
@@ -210,61 +203,157 @@ func (s *AuthServer) validateGithubAuthCallback(q url.Values) (*GithubAuthRespon
...
+
+	// traits is the list of traits for this user.
+	traits map[string][]string
+
+	// sessionTTL is how long this session will last.
+	sessionTTL time.Duration
+}
+
+func (s *AuthServer) calculateGithubUser(connector services.GithubConnector, claims *services.GithubClaims, request *services.GithubAuthRequest) (*createUserParams, error) {
+	var p createUserParams
```
Comment (klizhentas) [inline lib/auth/github.go]: why not:

```go
p := createUserParams{
     connectorName: connector.GetName(),
     username: claims.Username,
}
```

File: lib/auth/github.go [hunk 5]
```diff
@@ -210,61 +203,157 @@ func (s *AuthServer) validateGithubAuthCallback(q url.Values) (*GithubAuthRespon
...
 	}
+	p.roles = modules.GetModules().RolesFromLogins(p.logins)
+	p.traits = modules.GetModules().TraitsFromLogins(p.logins, p.kubeGroups)
+
+	// Pick smaller for role: session TTL from role or requested TTL.
+	roles, err := services.FetchRoles(p.roles, s.Access, p.traits)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	roleTTL := roles.AdjustSessionTTL(defaults.CertDuration)
```
Comment (klizhentas) [inline lib/auth/github.go]: should be Max duration, not default cert duration, so we can process larger than default values

File: lib/auth/oidc.go [hunk 6]
```diff
@@ -325,52 +314,71 @@ func claimsToTraitMap(claims jose.Claims) map[string][]string {
 	return traits
 }
 
-func (a *AuthServer) createOIDCUser(connector services.OIDCConnector, ident *oidc.Identity, claims jose.Claims) error {
-	roles, err := a.buildOIDCRoles(connector, claims)
+func (a *AuthServer) calculateOIDCUser(connector services.OIDCConnector, claims jose.Claims, ident *oidc.Identity, request *services.OIDCAuthRequest) (*createUserParams, error) {
+	var err error
+	var p createUserParams
```
Comment (klizhentas) [inline lib/auth/oidc.go]: ditto as other user

File: lib/auth/saml.go [hunk 7]
```diff
@@ -105,47 +120,74 @@ func assertionsToTraitMap(assertionInfo saml2.AssertionInfo) map[string][]string
 	return traits
 }
 
-func (a *AuthServer) createSAMLUser(connector services.SAMLConnector, assertionInfo saml2.AssertionInfo, expiresAt time.Time) error {
-	roles, err := a.buildSAMLRoles(connector, assertionInfo)
+func (a *AuthServer) calculateSAMLUser(connector services.SAMLConnector, assertionInfo saml2.AssertionInfo, request *services.SAMLAuthRequest) (*createUserParams, error) {
+	var err error
+	var p createUserParams
```
Comment (klizhentas) [inline lib/auth/saml.go]: ditto

**Review by klizhentas — CHANGES_REQUESTED** [review]
still have a couple of comments and one actual change

**Review by klizhentas — APPROVED** [review]

**Pull request merged**